File: Editorial/44_Wildcard_Matching/CSharp/1_Recursion_with_Memoization.cs
public class Solution {
    Dictionary<string, bool> dp = new Dictionary<string, bool>();
    string s;
    string p;

    string remove_duplicate_stars(string p) {
        var new_string = new StringBuilder();
        foreach (var c in p) {
            if (new_string.Length == 0 || c != '*')
                new_string.Append(c);
            else if (new_string[new_string.Length - 1] != '*')
                new_string.Append(c);
        }

        return new_string.ToString();
    }

    bool helper(int si, int pi) {
        var key = si + "," + pi;
        if (dp.ContainsKey(key))
            return dp[key];
        if (pi == p.Length)
            dp[key] = si == s.Length;
        else if (si == s.Length)
            dp[key] = pi + 1 == p.Length && p[pi] == '*';
        else if (p[pi] == s[si] || p[pi] == '?')
            dp[key] = helper(si + 1, pi + 1);
        else if (p[pi] == '*')
            dp[key] = helper(si, pi + 1) || helper(si + 1, pi);
        else
            dp[key] = false;
        return dp[key];
    }

    public bool IsMatch(string s, string p) {
        dp.Clear();
        this.s = s;
        this.p = remove_duplicate_stars(p);
        return helper(0, 0);
    }
}

File: Editorial/44_Wildcard_Matching/JavaScript/1_Recursion_with_Memoization.js
var isMatch = function (s, p) {
    var dp = {};
    var remove_duplicate_stars = function (p) {
        var new_string = "";
        for (var c of p) {
            if (new_string.length == 0 || c != "*") new_string += c;
            else if (new_string[new_string.length - 1] != "*") new_string += c;
        }
        return new_string;
    };
    var helper = function (si, pi) {
        var key = si + "," + pi;
        if (key in dp) return dp[key];
        if (pi == p.length) dp[key] = si == s.length;
        else if (si == s.length) dp[key] = pi + 1 == p.length && p[pi] == "*";
        else if (p[pi] == s[si] || p[pi] == "?")
            dp[key] = helper(si + 1, pi + 1);
        else if (p[pi] == "*")
            dp[key] = helper(si, pi + 1) || helper(si + 1, pi);
        else dp[key] = false;
        return dp[key];
    };
    p = remove_duplicate_stars(p);
    return helper(0, 0);
};
